frame/offences/src/lib.rs

```diff
@@ -26,42 +26,68 @@ pub mod migration;
 mod mock;
 mod tests;
 
-use codec::{Decode, Encode};
-use frame_support::weights::Weight;
-use sp_runtime::{traits::Hash, Perbill};
-use sp_staking::{
-	offence::{Kind, Offence, OffenceDetails, OffenceError, OnOffenceHandler, ReportOffence},
-	SessionIndex,
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{weights::Weight, BoundedVec};
+use sp_runtime::traits::Hash;
+use sp_staking::offence::{
+	Kind, MaxOffenders, MaxReporters, Offence, OffenceDetails, OffenceError, OnOffenceHandler,
+	ReportOffence,
 };
 use sp_std::prelude::*;
 
 pub use pallet::*;
 
 /// A binary blob which represents a SCALE codec-encoded `O::TimeSlot`.
-type OpaqueTimeSlot = Vec<u8>;
+///
+/// Needs to be bounded since it is stored in `ReportIndexStorage`.
+type OpaqueTimeSlotOf<T> = BoundedVec<u8, <T as Config>::MaxOpaqueTimeSlotEncodedLen>;
 
 /// A type alias for a report identifier.
 type ReportIdOf<T> = <T as frame_system::Config>::Hash;
 
+type SameKindReportsOf<T, TimeSlot> =
+	BoundedVec<(TimeSlot, ReportIdOf<T>), <T as Config>::MaxSameKindReports>;
+
+type ConcurrentReportsOf<T> = BoundedVec<ReportIdOf<T>, <T as Config>::MaxConcurrentReportsPerTime>;
+
 #[frame_support::pallet]
 pub mod pallet {
 	use super::*;
 	use frame_support::pallet_prelude::*;
 
 	#[pallet::pallet]
 	#[pallet::generate_store(pub(super) trait Store)]
-	#[pallet::without_storage_info]
 	pub struct Pallet<T>(_);
 
 	/// The pallet's config trait.
 	#[pallet::config]
 	pub trait Config: frame_system::Config {
 		/// The overarching event type.
-		type Event: From<Event> + IsType<<Self as frame_system::Config>::Event>;
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
 		/// Full identification of the validator.
-		type IdentificationTuple: Parameter;
+		type IdentificationTuple: Parameter + MaxEncodedLen;
 		/// A handler called for every offence report.
 		type OnOffenceHandler: OnOffenceHandler<Self::AccountId, Self::IdentificationTuple, Weight>;
+
+		/// Maximum number of reports of the same kind that happened at a specific time slot.
+		#[pallet::constant]
```

Comment:
Do all of these really need to be in metadata? harmless, but I wonder if you had thought about each.